### PR TITLE
Clang-cl refuses to convert the types a private base-class

### PR DESCRIPTION
**Pull request by bfierz**
Using clang-cl to compile a project using std_allocator results in the following error:

memory\include\foonathan/memory/std_allocator.hpp(137,31): error : cannot cast 'const std::_Wrap_alloc<foonathan::memory::std_allocator<int, foonathan::memory::temporary_allocator, std::mutex> >' to its private base class 'const foonathan::memory::allocator_storage<foonathan::memory::reference_storage<foonathan::memory::temporary_allocator>, std::mutex>'
            : alloc_reference(alloc)

In order to fix this issue, the inhertiance was made public. I am not particularly happy with the fix, as it seems to be very specific to the combination of clang-cl with the STL of MSVC toolchan v140 (VS 2015).

### CODE REVIEW

**Comment by foonathan** [comment]
Don't worry about CI failure, that's my fault.

If this is a hack, could you ifdef it and add a comment linking to this PR?

**Comment by bfierz** [comment]
I've tried to narrow down the bug. It seems to be very specific to non-optimized builds using clang-cl. I was able to build a minimum example, but didn't manage to find a different work-around:
https://godbolt.org/g/oA6Fh4
Unfortunatelly, it can only be reproduced locally, as Godbolts compiler explorere doesn't support clang-cl.
If you're ok with it, I'm trying to add clang-cl to the appveyor configuration. I'm using it already in my own project (vcl on github).

File: include/foonathan/memory/std_allocator.hpp [hunk 1]
```diff
@@ -74,7 +74,12 @@ namespace foonathan
         /// synchronization is done via a \c Mutex which defaults to \ref default_mutex.
         /// \ingroup memory adapter
         template <typename T, class RawAllocator, class Mutex = default_mutex>
-        class std_allocator : FOONATHAN_EBO(allocator_reference<RawAllocator, Mutex>)
+        class std_allocator :
+#if defined _MSC_VER && !defined __clang__
```
Comment (foonathan) [inline include/foonathan/memory/std_allocator.hpp]: The logic seems incorrect, wouldn't you also want this for non MSVC compilers?

Comment (bfierz) [reply include/foonathan/memory/std_allocator.hpp]: That's embarrassing. You're absolutely right. My initial formulation was different, and I messed it up.

**Comment by foonathan** [comment]
Thank you, especially for the CI.

**Pull request merged**